### PR TITLE
Create ScrewTheoryLib

### PR DESCRIPTION
**Pull request by PeterBowman**
This PR adds a [Screw Theory](https://en.wikipedia.org/wiki/Screw_theory) inverse kinematics solver and related utilities with a broader goal in mind (https://github.com/roboticslab-uc3m/kinematics-dynamics/issues/119).

The core of this implementation relies on a set of solved geometric problems, among which we find the three well-established [Paden-Kahan subproblems](https://en.wikipedia.org/wiki/Paden-Kahan_subproblems) as well as a complementary collection of four additional cases as described in José Manuel Pardos-Gotor. *Screw Theory for Robotics - A practical approach for Modern Robot KINEMATICS - An Illustrated Handbook.* Amazon Fulfillment, 9 2018, ISBN 978-1717931818.

KDL has been chosen as the framework that provides all necessary abstractions for such geometric primitives as HT frames, vectors, joint arrays, etc. KDL wrappers for FK/IK solvers are provided as well.

New entities in the `roboticslab::` namespace:

* **PoeExpression:** [product of exponentiales (POE) formula](https://en.wikipedia.org/wiki/Product_of_exponentials_formula). Helper methods are provided to seamlessly convert to and from a `KDL::Chain`.

* **MatrixExponential:** each term of a POE formula ([ref](https://en.wikipedia.org/wiki/Matrix_exponential)).

* **ScrewTheoryIkSubproblem:** interface for all available (seven so far) IK subproblems.

* **ScrewTheoryIkProblem:** provides a `solve()` method to compute IK for a target pose.

* **ScrewTheoryIkProblemBuilder:** decouples solver initialization from runtime operations. This class searches for a valid sequence of local geometric subproblems that can solve a particular global IK problem. If found, an instance of *ScrewTheoryIkProblem* is created.

Non-exhaustive tests have proven that the Screw Theory algorithm as implemented here performs 5-10 times faster than a numeric Newton-Raphson solver provided by KDL. All available solutions for TEO's right arm are found in ~0.05 milliseconds.

Sample usage:

```c++
PoeExpression poe; // already defined and populated
ScrewTheoryIkProblemBuilder builder(poe);
ScrewTheoryIkProblem * ikProblem = builder.build();

KDL::Frame H;
ScrewTheoryIkProblem::Solutions solutions;

if (ikProblem->solve(H, solutions)
{
    // iterate over all available solutions
}
```

TODOs not covered here:

* out-of-reach poses not thoroughly tested 
* simple API that could be improved upon
* configuration selector, i.e. pick the best solution available (cf. [`AsibotConfiguration`](https://github.com/roboticslab-uc3m/kinematics-dynamics/blob/6ed8110a8a5879d9d119c2312256e154af73f880/libraries/YarpPlugins/AsibotSolver/AsibotConfiguration.hpp))

### CODE REVIEW

File: doc/Doxyfile [hunk 1]
```diff
@@ -593,7 +593,7 @@ LAYOUT_FILE            =
 # of the bibliography can be controlled using LATEX_BIB_STYLE. To use this
 # feature you need bibtex and perl available in the search path.
 
-CITE_BIB_FILES         =
+CITE_BIB_FILES         = bibtex/pardosgotor2018str_handbook.bib
```
Comment (jgvictores) [inline doc/Doxyfile]: Maybe we could rename this file to something more general?

We should also cite Khatib some day :smile:

**Review by jgvictores — CHANGES_REQUESTED** [review]
Looks really amazing. I basically pointed just one single thing out, regarding the `.bib` file name.

File: doc/Doxyfile (hunk 1 above)
Comment (PeterBowman) [reply doc/Doxyfile]: I followed the pattern 1 .bib = 1 cite per [asibot-main/doc/bibtex](https://github.com/roboticslab-uc3m/asibot-main/tree/8203df6558a1eb0582a1d3c94ca7294edf83571f/doc/bibtex). How would you like it to be?

Comment (jgvictores) [reply doc/Doxyfile]: - Single file: I'm currently using Mendeley as a reference manager, and tend to export whatever is in a group for a certain purpose ot a single `.bib` file. When you add a new reference, you just update this file.
- Separate files: Cleaner, but you have to remember to update the `Doxyfile`. You might be inclined to think it automatically parses the entire directory, it's a pitty it doesn't.

Maybe slightly more inclined towards the current `1 .bib = 1 cite` pattern after writing/reading my own reasons. What is your current opinion?

Comment (PeterBowman) [reply doc/Doxyfile]: Perhaps a single .bib file to keep things tidy, but move it out of the `bibtex` directory, then. Not having a strong opinion on this anyway.

Comment (jgvictores) [reply doc/Doxyfile]: I'd be fine with the single `.bib` file too. `/doc/kinematics-dynamics.bib` FTW then?

Comment (PeterBowman) [reply doc/Doxyfile]: Done!

**Review by jgvictores — APPROVED** [review]

**Comment by PeterBowman** [comment]
Follow-up: https://github.com/roboticslab-uc3m/kinematics-dynamics/commit/e8ed37ccf0b866016b0683673ac9bbce3131cec3 (some renaming + ISBN).

**Pull request merged**